### PR TITLE
Diamond search for half resolution motion estimation

### PR DESCRIPTION
**Pull request by magwyz**
It currently does not use previous frames computed motion.

### CODE REVIEW

**Comment by magwyz** [comment]
This seems to reduce the compression time. Here are the AWCY runs:
https://beta.arewecompressedyet.com/?job=diamond_me_hres_ref%402019-03-01T16%3A36%3A57.994Z&job=diamond_me_hres_test%402019-03-01T16%3A36%3A29.656Z

**Comment by tdaede** [comment]
Looks like an acceptable loss to take all the time.

File: src/encoder.rs [hunk 1]
```diff
@@ -1106,6 +1106,22 @@ pub fn motion_compensate<T: Pixel>(
   }
 }
 
+pub fn save_block_motion<T: Pixel>(
+   fi: &FrameInvariants<T>, fs: &mut FrameState<T>,
```
Comment (lu-zero) [inline src/encoder.rs]: Maybe take in directly `w_in_b` and `h_in_b`, as you did for `get_subset_predictors`?

File: src/partition.rs [hunk 2]
```diff
@@ -761,6 +761,10 @@ impl MotionVector {
       col: (self.col / 8) * 8
     }
   }
+
+  pub fn is_null(self) -> bool {
```
Comment (lu-zero) [inline src/partition.rs]: `is_zero()` would be less confusing :)

File: src/me.rs [hunk 3]
```diff
@@ -746,29 +758,53 @@ pub fn estimate_motion_ss2<T: Pixel>(
...
+            fi, &po,
+            &fs.input_hres, &rec.input_hres,
+            &predictors, fi.sequence.bit_depth,
+            global_mv, lambda,
+            mvx_min >> 1, mvx_max >> 1, mvy_min >> 1, mvy_max >> 1,
+            blk_w >> 1, blk_h >> 1,
+            &mut best_mv, &mut lowest_cost,
+            &mut None, 0
+          );
+        } else {
```
Comment (lu-zero) [inline src/me.rs]: Make them two functions and push the branch as up as possible in the call tree.

**Comment by magwyz** [comment]
Here are the latest AWCY runs:
https://beta.arewecompressedyet.com/?job=diamond_me_hres_ref%402019-03-14T16%3A45%3A14.540Z&job=diamond_me_hres_test

File: src/me.rs (hunk 3 above)
Comment (magwyz) [reply src/me.rs]: I extracted the if clause from the loop in the last version of this PR.
I have no issue making two functions. I however do not see in practice how to do it properly.

**Comment by lu-zero** [comment]
I'd rework it to use the same approach used in #1118 :)

File: src/me.rs [hunk 4]
```diff
@@ -731,91 +731,141 @@ pub fn estimate_motion_ss4<T: Pixel>(
...
+    frame_mvs: &[MotionVector], frame_ref_opt: &Option<Arc<ReferenceFrame<T>>>,
+    po: PlaneOffset, rec: &Arc<ReferenceFrame<T>>,
+    global_mv: &[MotionVector; 2], lambda: u32,
+    mvx_min: isize, mvx_max: isize, mvy_min: isize, mvy_max: isize,
+    blk_w: usize, blk_h: usize,
+    best_mv: &mut MotionVector, lowest_cost: &mut u64
+  );
+}
+
+pub struct DiamondSearchSS2 {}
```
Comment (lu-zero) [inline src/me.rs]: We could keep all in a single trait and struct placeholder maybe?

Comment (magwyz) [reply src/me.rs]: Sure.
This is in the last version of the PR.

**Comment by lu-zero** [comment]
It needs a rebase now =/

**Comment by magwyz** [comment]
I've just done the rebase.

File: src/me.rs [hunk 5]
```diff
@@ -397,15 +452,15 @@ pub struct FullSearch {}
...
   fn full_pixel_me<T: Pixel>(
     fi: &FrameInvariants<T>, fs: &FrameState<T>, rec: &Arc<ReferenceFrame<T>>,
-    po: &PlaneOffset, bo: &BlockOffset, lambda: u32, ref_slot: usize,
+    po: &PlaneOffset, bo: &BlockOffset, lambda: u32,
     cmv: MotionVector, pmv: [MotionVector; 2], mvx_min: isize, mvx_max: isize,
     mvy_min: isize, mvy_max: isize, blk_w: usize, blk_h: usize,
     best_mv: &mut MotionVector, lowest_cost: &mut u64, ref_frame: usize
   ) {
-    let frame_mvs = &fs.frame_mvs[ref_slot];
+    let frame_mvs = &fs.frame_mvs[ref_frame - LAST_FRAME];
```
Comment (tdaede) [inline src/me.rs]: I know this pattern is from libaom.... but I hate it :)

I'll change the whole codebase in my own patch.

**Pull request merged**